crates/sui-types/src/base_types.rs

```diff
@@ -128,6 +131,89 @@ impl From<&ObjectInfo> for ObjectRef {
     }
 }
 
+#[derive(Clone, Serialize, Deserialize, JsonSchema, Ord, PartialOrd, Eq, PartialEq, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct DynamicFieldInfo {
+    pub name: String,
+    pub type_: DynamicFieldType,
+    pub object_type: String,
+    pub object_id: ObjectID,
+    pub version: SequenceNumber,
+    pub digest: ObjectDigest,
+}
+
+impl DynamicFieldInfo {
+    pub fn new(
+        oref: &ObjectRef,
+        o: &Object,
+        get_object_info: &dyn Fn(&ObjectID) -> Option<(StructTag, SequenceNumber, ObjectDigest)>,
+    ) -> Option<Self> {
+        let field = o.data.try_as_move()?;
+        if !is_dynamic_field(&field.type_) {
+            return None;
+        }
+        // Assuming dynamic field's bytearray structured as following [field object id][field name][1][object id]
+        // field object id and object id is the same for dynamic field, and different for dynamic object.
+
+        // index of the 1u8 byte
+        let (index, _) = field
+            .contents()
+            .iter()
+            .find_position(|byte| **byte == 1u8)?;
+        let name = &field.contents()[ObjectID::LENGTH..index];
+
+        let name = if name.len() == ObjectID::LENGTH {
+            ObjectID::try_from(name).ok()?.to_hex_literal()
+        } else if let Ok(name) = String::from_utf8(name.to_vec()) {
+            name
+        } else {
+            Base64::encode(name)
+        };
```

Comment:
This is likely wrong, need to find a way to safely parse the object ids and field name, maybe use TypeTag information from object type